runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -3363,3 +3371,103 @@ J9::ValuePropagation::innerConstrainAcall(TR::Node *node)
 
    return node;
    }
+
+TR_OpaqueClassBlock *
+J9::ValuePropagation::findLikelySubtype(TR_OpaqueClassBlock *klass)
+   {
+   if (!klass || TR::VPConstraint::isSpecialClass((uintptr_t)klass) || comp()->compileRelocatableCode())
+      return NULL;
+
+   int32_t numDims = 0;
+   klass = ((TR_J9VM *)fe())->getBaseComponentClass(klass, numDims);
+
+   // If the returned base component class is still an array, it's a primitive array
+   if (TR::Compiler->cls.isClassArray(comp(), klass))
+      return NULL; // primitive array, should be fixed-type anyway
+
+   if (TR::Compiler->cls.isInterfaceClass(comp(), klass)
+       || TR::Compiler->cls.isAbstractClass(comp(), klass))
+      {
+      TR::ClassTableCriticalSection lock(fe());
+      auto *chTable = comp()->getPersistentInfo()->getPersistentCHTable();
+      klass = chTable->findSingleConcreteSubClass(klass, comp());
+      }
+   // If it's a concrete class and it has been extended, there is no particularly likely type
```

Comment:
Is it intentional that a startup phase check is there in one side of this if but not in the other ? i.e. do we need to check for it even when abstract or interface class is involved ? i.e. I assume the reason is that the class hierarchy may not have stabilized, but that sounds like we should be consistent either way.

Reply:
That's a good question. I suppose we could check if it's the start up phase for the case of abstract or interface classes to be consistent. I looked at other places where `findSingleConcreteSubClass` is used, we never seem to concern if it's in the start up phase or not. Then we could also argue that we might just remove the startup phase check for the concrete class case to be consistent. @jdmpapin I'm wondering what you think

Reply:
The current state of this PR is consistent in a certain way with `invokeinterface` inlining, which always uses the single concrete subtype of an interface (if there is one), but which, in deciding the guard test to use, only bets that a non-extended class will remain non-extended after the startup phase is over.

That said, I don't have a strong opinion on which cases (if any) should be subject to the startup phase condition. I don't think it's possible to determine that just by reasoning. Our decision here won't affect the performance of anything yet, and once it does, the effects will initially be limited to workloads using arrays of value types. So we have a lot of leeway to just decide more or less arbitrarily, and then adjust later if/when needed.

Reply:
Okay thanks, I see what you mean.